patchwork-registries/src/main/java/net/minecraftforge/registries/ForgeRegistry.java

```diff
@@ -40,17 +41,68 @@
 import net.minecraft.util.registry.DefaultedRegistry;
 import net.minecraft.util.registry.Registry;
 
-public class ForgeRegistry<V extends IForgeRegistryEntry<V>> implements IForgeRegistry<V> {
+import net.patchworkmc.impl.registries.ExtendedVanillaRegistry;
+
+public class ForgeRegistry<V extends IForgeRegistryEntry<V>> implements IForgeRegistryModifiable<V>, IForgeRegistryInternal<V> {
 	public static Marker REGISTRIES = MarkerManager.getMarker("REGISTRIES");
 	private static Logger LOGGER = LogManager.getLogger();
-	private Identifier name;
-	private Registry<V> vanilla;
-	private Class<V> superType;
+	private final Identifier name;
+	private final boolean isVanilla;
+	private final Registry<V> vanilla;
+	private final Class<V> superType;
+	private final Map<Identifier, ?> slaves = Maps.newHashMap();
+	private final CreateCallback<V> create;
+	private final AddCallback<V> add;
+	private final ClearCallback<V> clear;
+	private final RegistryManager stage;
+	private final boolean isModifiable;
+
+	private boolean isFrozen = false;
+
+	/**
+	 * Called by RegistryBuilder, for moded registries.
+	 * @param stage
+	 * @param name
+	 * @param builder
+	 */
+	public ForgeRegistry(RegistryManager stage, Identifier name, RegistryBuilder<V> builder) {
+		this.stage = stage;
+		this.name = name;
+		this.superType = builder.getType();
+		this.create = builder.getCreate();
+		this.add = builder.getAdd();
+		this.clear = builder.getClear();
+		this.isModifiable = builder.getAllowModifications();
 
+		this.vanilla = new ExtendedVanillaRegistry<>(this);
+		this.isVanilla = false;
+
+		if (this.create != null) {
+			this.create.onCreate(this, stage);
+		}
+	}
+
+	/**
+	 * For vanilla registeries.
+	 * @param name
+	 * @param vanilla
+	 * @param superType
+	 */
 	public ForgeRegistry(Identifier name, Registry<V> vanilla, Class<V> superType) {
 		this.name = name;
 		this.vanilla = vanilla;
 		this.superType = superType;
+
+		this.create = null;
+		this.add = null;
```

Comment:
What would these callbacks be used for?

Reply:
These callbacks are set in the RegistryBuilder, when the ForgeRegistry is created. We should keep this to maintain compatibility with Forge

Reply:
This should be fired then.
I believe coderbot's comment on RegistryEntryAddedCallback would resolve this.

Reply:
In the latest commit, the mixin injection is replaced by a RegistryEntryAddedCallback